### PR TITLE
Initial support for pmi based asm diffs

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Add jit-pmi-dasm as a new tool. Add plumbing for
this tool to jit-diffs as `pmidiff`, so one can now say:

  jit-diff pmidiff  --base --base_root d:\repos\coreclr2 --diff

There are some crufty parts. Because we're not (yet?) using altjits
for same-arch pmi, we have to copy the jit under test into the
test overlay dir. This means carefully saving off the jit that was
there and putting it back when done, as the primary jit name can't
be altered. Also we can't run base and diff jit tasks at the same time.
So we run one group, then the other.

Update pmi to newline after new method banner so the jit dasm output
starts on its own line.

Output is compatible with existing jit-analyze.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

Have not tested with altjits and such, or on non-windows. So there may be issues to sort out.

Here's a corelib diff from my recent inlining change, via PMI:
```
Total bytes of diff: -72 (-0.01% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
         -72 : System.Private.CoreLib.dasm (-0.01% of base)
1 total files with size differences (1 improved, 0 regressed), 0 unchanged.
Top method improvements by size (bytes):
         -36 : System.Private.CoreLib.dasm - ReadOnlySpan`1:GetPinnableReference():byref:this (2 methods)
         -36 : System.Private.CoreLib.dasm - Span`1:GetPinnableReference():byref:this (2 methods)
2 total methods with size differences (2 improved, 0 regressed), 2204 unchanged.
```
and asm diff:
```diff
PREPALL type# 352 method# 5945 System.ReadOnlySpan`1[[System.Double, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::GetPinnableReference
 ;
 ;  V00 this         [V00,T00] (  4,  3.50)   byref  ->  rcx         this
 ;* V01 loc0         [V01    ] (  0,  0   )  struct ( 8) zero-ref    ld-addr-op
-;  V02 OutArgs      [V02    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
-;  V03 tmp1         [V03,T01] (  2,  1   )   byref  ->  rax         V01._value(offs=0x00) P-INDEP
+;# V02 OutArgs      [V02    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
+;* V03 tmp1         [V03    ] (  0,  0   )    long  ->  zero-ref   
+;  V04 tmp2         [V04,T01] (  2,  1   )   byref  ->  rax         V01._value(offs=0x00) P-INDEP
 ;
-; Lcl frame size = 40
+; Lcl frame size = 0
 G_M32161_IG01:
-       sub      rsp, 40
 G_M32161_IG02:
        cmp      dword ptr [rcx+8], 0
        jne      SHORT G_M32161_IG04
-       xor      rcx, rcx
-       call     Unsafe:AsRef(long):byref
-       nop      
+       xor      rax, rax
 G_M32161_IG03:
-       add      rsp, 40
        ret      
 G_M32161_IG04:
        mov      rax, bword ptr [rcx]
 G_M32161_IG05:
-       add      rsp, 40
        ret      
-; Total bytes of code 31, prolog size 4 for method ReadOnlySpan`1:GetPinnableReference():byref:this
+; Total bytes of code 13, prolog size 0 for method ReadOnlySpan`1:GetPinnableReference():byref:this
```

Takes advantage of dotnet/coreclr#18306 but does not require it (that change will just reduce volume of no-diff methods that always get disassembled as they are part of PMI proper or the jitted cohort in dependent assemblies).

We should consider disabling zap/r2r for PMI but I will leave that for later if it seems valuable.

**Comment by AndyAyersMS** [comment]
Concurrent operation still not working right. Need to tweak the layering a bit.

**Comment by AndyAyersMS** [comment]
At least two issues:
* jit-dasm-pmi can't own install/restore of the jit if it's invoked via jit-diff, as the latter will make many concurrent invocations of jit-dasm-pmi. I moved the copy higher up but need to rework what should happen if jit-dasm-pmi is used on its own.
* the `NextMethodToPrep.marker` files written by PMI need to be distinct per process, or just suppressed when doing diff runs. Otherwise the concurrent pmi processes (kicked off by the concurrent jit-dasm-pmis) collide trying to update this file.

**Comment by BruceForstall** [comment]
I suppose there's a related question to your issue no. 2: can/should we make PMI (optionally) run in parallel, on a single assembly? And/or, should PMI and/or a PMI driver run it in parallel on separate assemblies, when given an assembly set or directory tree?

**Comment by AndyAyersMS** [comment]
With the latest commit I can now run via jit-diffs over all fx assemblies.
```
D:\repos\coreclr>jit-diff pmidiff  --base --base_root d:\repos\coreclr2 --diff -f -t ABC
Using --output=D:\repos\coreclr\bin\diffs
Using --base=d:\repos\coreclr2\bin\Product\Windows_NT.x64.Checked
Using --diff=D:\repos\coreclr\bin\Product\Windows_NT.x64.Checked
Using --core_root=D:\repos\coreclr\bin\tests\Windows_NT.x64.Release\Tests\Core_Root
Beginning PMI Diffs for System.Private.CoreLib.dll, framework assemblies
Completed PMI Diffs for System.Private.CoreLib.dll, framework assemblies in 466.70s
Analyzing diffs...
Found files with textual diffs.
Summary:
(Note: Lower is better)
Total bytes of diff: -72 (0.00% of base)
    diff is an improvement.
Total byte diff includes 0 bytes from reconciling methods
        Base had    0 unique methods,        0 unique bytes
        Diff had    0 unique methods,        0 unique bytes
Top file improvements by size (bytes):
         -72 : System.Private.CoreLib.dasm (-0.01% of base)
1 total files with size differences (1 improved, 0 regressed), 129 unchanged.
Top method improvements by size (bytes):
         -36 : System.Private.CoreLib.dasm - ReadOnlySpan`1:GetPinnableReference():byref:this (2 methods)
         -36 : System.Private.CoreLib.dasm - Span`1:GetPinnableReference():byref:this (2 methods)
2 total methods with size differences (2 improved, 0 regressed), 169790 unchanged.
Completed analysis in 134.85s
```

We might want to tone down the internal verbosity of PMI as there are many methods passed over (because they have been prejitted) so lots of entries of the form:
```
Start type <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]
PREPALL type# 0 method# 0 <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::get_message
Completed method <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::get_message, elapsed ms: 0.01
PREPALL type# 0 method# 1 <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::Equals
Completed method <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::Equals, elapsed ms: 0.00
PREPALL type# 0 method# 2 <>f__AnonymousType0`1[[System.Object, System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]::GetHashCode
```
Also would be nice to dump generic type names without fully verbose assembly IDs for type parameters. Not sure yet how to do that.

And we probably want to suppress elapsed time when PMI is invoked via jit-diff-pmi as it just causes spurious non-code diffs.

**Comment by AndyAyersMS** [comment]
Long analysis times seem to be caused by two factors
* unnecessary chatter in pmi output
* we run the overall diff over entire directories, then if that sees a diff we analyze each pair of dasm files. Seems like instead this should work pairwise? That way analysis time should scale up with number of assemblies with diffs, not with number of assemblies.

**Comment by AndyAyersMS** [comment]
Ok, quieted down PMI a little (opt-in from jit-pmi-diff).

Think this is in reasonable shape now. Will do some non-windows testing shortly.

As for how to cope with parallelism -- yes this needs more thought. We have 3 levels of tooling in the "normal" workflow (jit-diff -> jit-dasm-pmi -> pmi) it seems painful to have all 3 know about orchestrating tasks in parallel, especially so with stateful things like needing to copy in the jit.

**Comment by AndyAyersMS** [comment]
Need a couple more fixes to make sure that `jit-diff diff` still works too. 

Most visible is that I needed to change PMI from a command to an option:
```
before:   jit-diff pmidiff --diff
after:    jit-diff diff --pmi --diff
```

**Comment by AndyAyersMS** [comment]
A few more usability tweaks. I've run this now on a full test tree (see example in dotnet/coreclr#18348).

Also some testing on OSX, though not full success there, yet as the jit asserts when disassembling an early-jitted method.

File: src/pmi/pmi.cs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.ms
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/pmi/pmi.cs]: Thanks, fixed.

File: src/pmi/pmi.cs [hunk 2]
```diff
@@ -316,64 +322,106 @@ class PrepareAll : PrepareBase
...
-        WriteAndFlushNextMethodToPrepMarker();
+        // For now only emit marker files if not in verbose mode.
+        //
+        // When may PMIs are invoked by jit-diff / jit-dasm-pmi we
+        // have trouble with file name contention as the names we
+        // pick are not unique enough. But the names also need to be
+        // predictable by DRIVEALL.
+        //
+        // So as a hack we rely on the fact that DRIVEALL always invokes
+        // PMI in verbose mode, and jit-dasm-pmi in quiet mode.
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/pmi/pmi.cs]: Seems reasonable. Though I will likely hold off for now and come back to this. Would like to get this version out so people can start using it...

Somewhere around here we may need to convert PMI over to using a more featureful command line parser. Unfortunately the one used by jit-diffs and friends seems to have lost favor and it's not clear if there's some common package that is actively maintained.

**Pull request merged**